### ISSUE TITLE
Rewrite Route related functionality to get rid of Route.NAMING_CONVENTION

### ISSUE THREAD

**Issue by denis-anisimov** [vaadin/flow#11915]
Ugly `Route.NAMING_CONVENTION` is an indicator which is used to use a hardcoded strategy to generate a navigation target path for a `@Route` component.

This functionality relies in the fragile assumption that nobody is going to use `Route.NAMING_CONVENTION` value as a real path for the route. Even though it's quite improbable situation such reliance smells.

I suggest to make an improvement and rewrite this :
**let's use `String[]` as a type for `value`**

That gives three wins immediately:
- Java allows to use _one single_ String instead if array with one element in the annotation values. It means that all the existing code will be backward compatible.
- The very first path will be used as usual : main route. All other will be used as aliases. That allows to reduce `@RouteAlias` annotation usage: in most cases you don't need different configs for aliases but instead just list paths as aliases and it's better to use one annotation with array value instead of bunch of `@RouteAlias`.
- An empty array of paths is a clear indication that no path are provided and this indicator can be used to auto generate the path instead of  `Route.NAMING_CONVENTION` indicator.

Similar issue is https://github.com/vaadin/flow/issues/4117.

**Comment by mvysny** [comment]
In Karibu-Testing, I have a set of tests compiled with Vaadin 14, which I then run with Vaadin 14, 20, 21 and 22 to make sure that Karibu-Testing works with all of those. The tests are using test routes to test navigation.

However, starting with Vaadin 22.0.0.alpha7 the routes are no longer binary-compatible: running the test suite will fail with
```
    java.lang.annotation.AnnotationTypeMismatchException: Incorrectly typed data found for annotation element public abstract java.lang.String[] com.vaadin.flow.router.Route.value() (Found data of type class java.lang.String[testing])
        at java.base/sun.reflect.annotation.AnnotationTypeMismatchExceptionProxy.generateException(AnnotationTypeMismatchExceptionProxy.java:57)
        at java.base/sun.reflect.annotation.AnnotationInvocationHandler.invoke(AnnotationInvocationHandler.java:86)
        at com.sun.proxy.$Proxy43.value(Unknown Source)
        at com.vaadin.flow.router.internal.RouteUtil.getAliasObjects(RouteUtil.java:371)
        at com.vaadin.flow.server.startup.AbstractRouteRegistryInitializer.checkForConflictingAnnotations(AbstractRouteRegistryInitializer.java:121)
```
From my point of view Vaadin 22 breaks binary compatibility with Vaadin 21 since Routes compiled with Vaadin 21 no longer works with Vaadin 22.

**Comment by mvysny** [comment]
@denis-anisimov I am facing a real issue: I have a set of utility classes for Vaadin and I'm trying to test them with multiple Vaadin versions. Those utilities also test Vaadin navigation, for which I need to define testing routes. And therein lies the problem. Yes it is a bit of a corner-case, but is causing me quite a headache.

**Comment by denis-anisimov** [comment]
I quite dislike new annotation introducing.
It's huge amount of code changes and dramatically drops maintainability : too many annotations which one has to care about.
Already now one should care about two annotations `Route` and `RouteAlias` carefully.
Third annotation makes the code very error-prone: it's extremely easy to introduce a mistake (and/or regression).

So 2 is not an option for me.

I would rather revert this change back than introduce a new annotation (new annotation also means changes in many add-ons).

I'm not convinced that this change brings serious problem. As I said it may be solved by introducing a new version of the library and since we are doing this for every add-on ( that was the point: we are already doing this and this is not an unresolvable problem) this is doable and requires just one extra version with only one dependency version change in the project config file (pom or whatever).
In case it introduce more issues (which I don't see at the moment) the change can be reverted.

**Comment by mstahv** [comment]
"Perfect is the greatest enemy of good". The change is kind of just polishing the internals, don't believe there will be ever an issue in a real project. Let's revert and return to this if there is some bigger renewal where we need to break the (binary) compatibility anyways.

**Comment by mvysny** [comment]
Thank you guys, the revert is landed in 22.0.0.beta1.

**Comment by pleku** [comment]
> Thank you guys, the revert is landed in 22.0.0.beta1.

Yes this is correct. Thanks @mvysny for bringing it up when you noticed the issue!
